rustler/src/alloc.rs

```diff
@@ -0,0 +1,46 @@
+use std::alloc::{GlobalAlloc, Layout};
+
+const SIZEOF_USIZE: usize = std::mem::size_of::<usize>();
+const MAX_ALIGN: usize = 8;
+
+#[cfg(feature = "allocator")]
+#[global_allocator]
+static ALLOCATOR: EnifAllocator = EnifAllocator;
+
+/// Allocator implementation that forwards all allocation calls to Erlang's allocator. Allows the
+/// memory usage to be tracked by the BEAM.
+pub struct EnifAllocator;
+
+unsafe impl GlobalAlloc for EnifAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        if layout.align() > MAX_ALIGN {
+            // overallocate and store the original pointer in memory immediately before the aligned
+            // section
```

Comment:
Could you add information why this is necessary? Why do we need to overallocate?

Reply:
Because we only get 64bit aligned buffers from Erlang. So, if we require e.g. 512bit (8 byte) alignment, 32 bytes, we ask for a buffer that can hold 32 + 8 bytes, then we will always be able to return a buffer with the necessary layout, just shiftedcompared to the original pointer. Since we need to keep that one around for `enif_free`, we storw it in the same buffer, just before the returned pointer. That's another `sizeof::<usize>()` bytes, we need to allocate additionally.